bigqueue_flush_test.go

```diff
@@ -0,0 +1,274 @@
+			t.Fatalf("enqueue failed :: %v", err)
+		}
+	}
+
+	checkDirtiness(bq, t, true, [3]bool{false, true, true})
+	checkTimesCalled(bq, t, 1, [3]int{1, 1, 0})
+
+	clock.Advance(time.Second)
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkDirtiness(bq, t, false, [3]bool{false, false, false})
+	checkTimesCalled(bq, t, 2, [3]int{1, 2, 1})
+}
+
+func TestResetOfBothFlushIntervalStates(t *testing.T) {
+	testDir := path.Join(os.TempDir(), fmt.Sprintf("testdir_%d", rand.Intn(1000)))
+	createTestDir(t, testDir)
+	defer deleteTestDir(t, testDir)
+
+	flushElapsedDuration := time.Second
+	clock := clockwork.NewFakeClock()
+
+	bq, err := NewMmapQueue(testDir, SetFlushElapsedDuration(flushElapsedDuration),
+		SetFlushIntervalMutateOps(2),
+		setClock(clock))
+
+	if err != nil {
+		t.Fatalf("unable to get BigQueue: %v", err)
+	}
+	defer bq.Close()
+
+	msg := []byte("a")
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkMutateOps(bq, t, 1)
+	clock.Advance(flushElapsedDuration) // time interval reached, flush happens, resets both prevFlushTime, mutateOps
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkMutateOps(bq, t, 0)
+	prevFlushTime := (bq.(*MmapQueue)).prevFlushTime
+	if prevFlushTime != clock.Now() {
+		t.Fatalf("expected prevFlushTime: %v to be reset to clock's now: %v", prevFlushTime, clock.Now())
+	}
+
+	clock.Advance(flushElapsedDuration / 2) // time moves ahead
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	// flush after 2 mutateOps, causes mutateOps reset and time reset to Now()
+
+	checkMutateOps(bq, t, 0)
+	prevFlushTime = (bq.(*MmapQueue)).prevFlushTime
+	if prevFlushTime != clock.Now() {
+		t.Fatalf("expected prevFlushTime: %v to be reset to clock's now: %v", prevFlushTime, clock.Now())
+	}
+
+}
+
+type flushCountingFile struct {
+	mmap.File
+	timesCalled int
+}
+
+func (m *flushCountingFile) Flush(flags int) error {
+	m.timesCalled++
+	return m.File.Flush(flags)
+}
+
+func setupFlushCountFile(q Queue, t *testing.T) {
+	bq := q.(*MmapQueue)
+	for i := 0; i < bq.conf.maxInMemArenas; i++ {
+		arena, err := bq.am.getArena(i)
+		if err != nil {
+			t.Fatalf("test setup failed, could not load arena %d because: %v", i, err)
+		}
+		arena.File = &flushCountingFile{arena.File, 0}
+	}
+	bq.index.indexArena.File = &flushCountingFile{bq.index.indexArena.File, 0}
+}
+
+func checkMutateOps(q Queue, t *testing.T, expected int64) {
+	bq := q.(*MmapQueue)
+	if bq.mutateOpsSinceFlush != expected {
+		t.Fatalf("expected mutateOpsSinceFlush %d, got %d", expected, bq.mutateOpsSinceFlush)
+	}
+}
+
+// expects first three arenas to be in memory
+func checkTimesCalled(q Queue, t *testing.T, indexTimesCalled int, timesCalled [3]int) {
+	bq := q.(*MmapQueue)
+	for i, expected := range timesCalled {
```

Comment:
same here

Reply:
I'll change the test to explictly set the maxInMemArenas to 3 to avoid regressions in case we change the default to something less which would cause test failure since the test assertions here are specifically for first three arenas.